### DISCUSSION TITLE
性能问题，随着插入数据增多查询逐步变慢

### DISCUSSION THREAD

**Discussion by jercas** [milvus-io/milvus#17997] [Q&A and General discussion]
Hi，我现在使用的是v2.0.2单机版本；
1. 参数如下：
    _DIM = 64
    _NLIST = 2048
    _NPROBE = 32
    _INDEX = IVF_FLAT
    _VEC_NUM ≈ 0.3M
2. 场景如下：
  我们是流式场景，每隔一段时间会生成一份新的全量item embedding进行插入，线上不间断查询。
  因为参考文档Q&A v2.0.2版本，当前并没有完善的去重机制，milvus并不会根据primary_key去对应覆盖。
  所以我们的方案是，按插入时间戳区分，每次新的全量embedding生成就存入一个新的partition，并删除旧partition。
3. 问题：
  1）查询速度不如预期：在consistency_level="Strong"的query下，平均时间为200ms；
  2）随着运行延时逐渐增大：在将consistency_level改为"Session"后，平均时间降为10ms，但随着不断地新partition插入，旧partition删除，平均query时间逐渐增长；
  3）删除不彻底：在删除旧partition后，我们发现虽然Attu(Partitions)和milvus-cli(show query_segment)中num_rows是单个partition的条数是正常的，但是整个collection的entities看起来是所有曾插入过的partitions条数的总和，是否删除parttion并未实际生效？ 
  PS：实际query时基于的是num_rows还是entities的数量呢？

**Comment by yhmo** [comment]
“随着不断地新partition插入，旧partition删除，平均query时间逐渐增长” ----- 这可能是个performance问题，能否创建一个issue
“collection的entities看起来是所有曾插入过的partitions条数的总和，是否删除parttion并未实际生效” ------- 删除partition操作是生效了的，只是目前collection.num_entities是从元数据里取的，而删除操作没有去更新元数据，这是个已知问题，后面会修复。
“实际query时基于的是num_rows还是entities的数量呢” ------ query是基于entities来做的，不会受collection.num_entities的影响

**Reply by jercas** [reply]
感谢回答，请问如果我在一个已load collection内新建一个partition并删除旧partition后，是否需要对该collection重新建index并load？
1. 是否需要操作？
2. 删除后，对该collection的query就是基于新partition的？

**Reply by yhmo** [reply]
在milvus里，数据是分片(segment)管理的，每个partition可能有数个segment，每个segment都会**单独**建立一个索引。删除一个partition，它所拥有的segment以及这些segment的索引就会被删除(逻辑上删除了，真正把东西删光有一个滞后期)。被逻辑上删除之后，就查询不到了。
load()原则上是调用一次即可，调用一次之后，建好的索引会被自动加载，新增的数据会被自动加载，删除的数据会被自动释放。但是如果有drop_index()的操作，则必须调release()一下再重新load()。所以你删除/新增一个partition不需要再load。新增的partition也会被自动加载到cache，在cache里的都能被查询到。

**Reply by jercas** [reply]
好的感谢答疑~
PS：一周运行后，在数据量不变和上述参数的前提下，query时间从10ms -> 600ms，我这边会提个Issue，辛苦
